src/spatial/utils.jl

```diff
@@ -0,0 +1,219 @@
+"""
+A file with types, structs and functions for spatial simulations
+"""
+
+"""
+stores info for a spatial jump
+"""
+struct SpatialJump{J}
+    site::J #where the jump happens
+    index::Int #diffusion or reaction
+    target_site::J #target of diffusive hop
+end
+
+############ abstract spatial system struct ##################
+"""
+Contains all info about the topology of the system
+"""
+abstract type AbstractSpatialSystem end
+
+"""
+returns neighbors of site
+"""
+function neighbors end
+
+"""
+returns total number of sites
+"""
+function number_of_sites end
+
+"""
+return the number of neighbors of a site
+"""
+function num_neighbors end
+
+# TODO make a graph struct with connectivity list, compare with CartesianGrid
+
+################### CartesianGrid <: AbstractSpatialSystem ########################
+#TODO store the number of neighbors for each site or store all neighbors for each site.
+struct CartesianGrid <: AbstractSpatialSystem
+    dimension::Int # dimension of the grid, e.g. 1, 2 or 3
+    linear_size::Int #side length of the grid
+end
+
+dimension(grid::CartesianGrid) = grid.dimension
+
+number_of_sites(grid) = grid.linear_size^dimension(grid)
+
+"""
+check if the given site_id is a valid site in the grid
+"""
+is_site(grid,site_id) = site_id >= 1 && site_id <= number_of_sites(grid)
```

Comment:
This won't always work right. Think about a 2D, 10x10 grid. You can get location 11 if you calculate the right neighbor of site 10, which is not a valid location. (Or maybe you don't use this in places where that can happen?)

Reply:
Yes, you are right. I fixed it